components/style/traversal.rs

```diff
@@ -113,24 +132,113 @@ pub trait DomTraversalContext<N: TNode> {
+    ///
+    /// The skip_root parameter is used in Gecko to style newly-appended children
+    /// without restyling the parent.
+    fn pre_traverse(root: N::ConcreteElement, stylist: &Stylist, skip_root: bool)
+                    -> PreTraverseToken
+    {
+        // If we should skip the root, traverse unconditionally.
+        if skip_root {
+            return PreTraverseToken {
+                traverse: true,
+                skip_root: true,
+            };
+        }
+
+        // Expand the snapshot, if any. This is normally handled by the parent, so
+        // we need a special case for the root.
+        //
+        // Expanding snapshots here may create a LATER_SIBLINGS restyle hint, which
+        // we will drop on the floor. This is fine, because we don't traverse roots
+        // with siblings.
+        debug_assert!(root.next_sibling_element().is_none());
+        if let Some(mut data) = root.mutate_data() {
+            if let Some(r) = data.as_restyle_mut() {
+                let _later_siblings = r.expand_snapshot(root, stylist);
+            }
+        }
+
+        PreTraverseToken {
+            traverse: Self::node_needs_traversal(root.as_node()),
+            skip_root: false,
+        }
+    }
+
+    /// Returns true if traversal should visit a text node. The style system never
+    /// processes text nodes, but Servo overrides this to visit them for flow
+    /// construction when necessary.
+    fn text_node_needs_traversal(_: N) -> bool { false }
+
+    /// Returns true if traversal is needed for the given node and subtree.
+    fn node_needs_traversal(node: N) -> bool {
+        // Non-incremental layout visits every node.
+        if cfg!(feature = "servo") && opts::get().nonincremental_layout {
+            return true;
+        }
+
+        match node.as_element() {
+            None => Self::text_node_needs_traversal(node),
+            Some(el) => {
+                // If the dirty descendants bit is set, we need to traverse no
+                // matter what. Skip examining the ElementData.
+                if el.has_dirty_descendants() {
+                    return true;
+                }
+
+                // Check the element data. If it doesn't exist, we need to visit
+                // the element.
+                let data = match el.borrow_data() {
+                    Some(d) => d,
+                    None => return true,
+                };
+
+                // Check what kind of element data we have. If it's Initial or Persistent,
+                // we're done.
+                let restyle = match *data {
+                    ElementData::Initial(ref i) => return i.is_none(),
+                    ElementData::Persistent(_) => return false,
+                    ElementData::Restyle(ref r) => r,
+                };
+
+                // Check whether we have any selector matching or re-cascading to
+                // do in this subtree. Snapshots should already be expanded.
```

Comment:
nit: Put the "Snapshots should already be expanded" bit in the assertion message?